### ISSUE TITLE
Acitvation function relu terrible performance

### ISSUE THREAD

**Issue by Fhilde** [FluxML/Flux.jl#1537]
**Description**
I noticed while programming some VAE, that the performance could get terrible and noticed that it came from using the relu activation function. I could reproduce the bug even with a very simple autoencoder. I'm using Julia 1.5.3 and Flux v0.10.4 on a standard laptop(Ubuntu18) using cpu.

**Code**
    #only imports
    using Flux
    using Flux: @epochs,mse
    
    [...]
    #just gets Flux Dataset and flattens it to one dimension and uses Flux.DataLoader to create batches of size 128
    loaded_X = getFluxMNIST() 
    model = Chain(
                Dense(inputsize, 256),
                Dense(256, 10),
                Dense(10, 256),
                Dense(256, inputsize,relu)); 
    
    loss(x) = mse(model(x), x) / batchsize
    opt = ADAM(0.01)
    
    @time @epochs 1 Flux.Optimise.train!(loss, params(model), loaded_X, opt)
    

**Performance**

    time of a single epoch when I drop the relu activation in the last layer:
    263.191092 seconds (7.25 M allocations: 19.577 GiB, 0.44% gc time)
    
    with relu:
    3395.982462 seconds (100.54 G allocations: 1.481 TiB, 9.52% gc time)
    
**Questions**
    any ideas why performance is so vastly different? It might only happen if relu is used for the last layer?

**Comment by DhairyaLGandhi** [comment]
Could you try updating your Flux version? 0.10 is pretty old at this point.

I'd also check for the performance of the forward pass and backwards pass separately using `@btime`

**Comment by Fhilde** [comment]
> Could you try updating your Flux version? 0.10 is pretty old at this point.

> I'd also check for the performance of the forward pass and backwards pass separately using `@btime`

**Comment by Fhilde** [comment]
> Could you try updating your Flux version? 0.10 is pretty old at this point.
sure, isn't it the last stable version? At least I used updated it with Pkg.update("Flux") just before trying it. But I will try with 0.11.4.

> I'd also check for the performance of the forward pass and backwards pass separately using `@btime`

Is that possible with @epochs? Or should just make a for loop where i can check pullback individually? Because I just tried @btime for the version without relu and just got:
193.939 s (7245219 allocations: 19.58 GiB)

Thanks a lot for the **fast** supply, I'm still *very*new to Julia so if I'm making some obvious mistakes please let me know!

**Comment by darsnack** [comment]
You would do something like:
```julia
using BenchmarkTools

model = Chain(
            Dense(inputsize, 256),
            Dense(256, 10),
            Dense(10, 256),
            Dense(256, inputsize,relu)); 
loss(x, m) = mse(m(x), x) / batchsize
ps = params(model)

# check forward pass performance
@btime model($(rand(Float32, inputsize)))
@btime Flux.gradient(() -> loss($(rand(Float32, inputsize)), $model), $ps)
```

Flux doesn't do LTS releases like Julia itself. Any minor version (e.g. v0.11.X for any X) is "stable" in the sense that the API will be stable, and we don't intentionally release broken code ever. Going from v0.11.X to v0.12.X can include API changes, but we try to provide soft deprecation paths, so you will get a warning about using any function that changes.

**Comment by Fhilde** [comment]
Thank you very much for the help and guidance!
Seem like it's not a problem in Version v0.12.0. I checked with the above version from @darsnack 

**v0.10.4**:
*without relu*:
  2.119 μs (8 allocations: 5.19 KiB)  - forward 
  47.653 μs (77 allocations: 166.59 KiB) -forward and backward 
*with relu*:
  2.109 μs (8 allocations: 5.19 KiB)
  6.938 ms (95916 allocations: 1.65 MiB)

**v0.12.0**:
*witout relu*:
   2.146 μs (8 allocations: 5.19 KiB)
  161.039 μs (444 allocations: 105.17 KiB)
*with relu*:
   2.209 μs (8 allocations: 5.19 KiB)
  159.879 μs (444 allocations: 105.64 KiB)

Interestingly the sigmoid activation fucntion from v0.10.4 seems to have  been quite a bit faster than the current one!
